### PR TITLE
Correctly resolve relative local paths pointing to patch files from dependencies

### PR DESCRIPTION
**Pull request by rlmumford**
Update from PR #161

### CODE REVIEW

**Comment by stucki** [comment]
This patch does exactly what I need. Works great, thank you!

File: src/Patch.php [hunk 1]
```diff
@@ -39,6 +39,13 @@ class Patch implements \JsonSerializable
      */
     public $depth;
 
+    /**
+     * The package that provides the patch.
+     *
+     * @var \Composer\Package\PackageInterface $provider
```
Comment (EmiiKhaos) [inline src/Patch.php]: Code style: Import the namespace and don't reference the FQCN

File: src/Plugin/Patches.php [hunk 2]
```diff
@@ -281,6 +281,24 @@ public function checkPatches(Event $event)
         }
     }
 
+    /**
+     * Check whether a given path is relative.
+     *
+     * @param string $url
+     * @return bool
```
Comment (EmiiKhaos) [inline src/Plugin/Patches.php]: Code style: Missing empty line between `@param` and `@return`

File: src/Plugin/Patches.php [hunk 3]
```diff
@@ -281,6 +281,24 @@ public function checkPatches(Event $event)
...
+     */
+    protected function isRelativeUrl($url) {
+        if (parse_url($url, PHP_URL_SCHEME) != '') {
+            return FALSE;
```
Comment (EmiiKhaos) [inline src/Plugin/Patches.php]: Code style: lowercase `false`

File: src/Plugin/Patches.php [hunk 4]
```diff
@@ -281,6 +281,24 @@ public function checkPatches(Event $event)
...
+        }
+
+        if ($url[0] == '/') {
+            return FALSE;
```
Comment (EmiiKhaos) [inline src/Plugin/Patches.php]: Code style: lowercase `false`

File: src/Plugin/Patches.php [hunk 5]
```diff
@@ -281,6 +281,24 @@ public function checkPatches(Event $event)
...
+        }
+
+        return TRUE;
```
Comment (EmiiKhaos) [inline src/Plugin/Patches.php]: Code style: lowercase `true`

File: src/Plugin/Patches.php [hunk 6]
```diff
@@ -322,6 +340,16 @@ public function postInstall(PackageEvent $event)
                     null,
                     new PatchEvent(PatchEvents::PRE_PATCH_APPLY, $package, $patch->url, $patch->description)
                 );
+
+                if ($this->isRelativeUrl($patch->url) && !empty($patch->provider)) {
+                  if (!$manager->isPackageInstalled($localRepository, $patch->provider)) {
+                    $this->io->write(' - <info>Installing '.$patch->provider->getName().'</info> to ensure access to relative patches.');
+                    $manager->getInstaller($patch->provider->getType())->install($localRepository, $patch->provider);
+                  }
```
Comment (EmiiKhaos) [inline src/Plugin/Patches.php]: Code style: Missing empty line after `}`

File: src/Plugin/Patches.php [hunk 7]
```diff
@@ -322,6 +340,16 @@ public function postInstall(PackageEvent $event)
...
+                  $providing_install_path = $manager->getInstallPath($patch->provider);
+                  $patch->url = $providing_install_path.'/'.$patch->url;
```
Comment (EmiiKhaos) [inline src/Plugin/Patches.php]: Code style: Use lower camel case `$providingInstallPath`

File: src/Plugin/Patches.php [hunk 8]
```diff
@@ -322,6 +340,16 @@ public function postInstall(PackageEvent $event)
...
+                }
```
Comment (EmiiKhaos) [inline src/Plugin/Patches.php]: Code style: Indentation is 4 spaces

File: src/Resolvers/DependencyPatches.php [hunk 9]
```diff
@@ -34,8 +34,9 @@ public function resolve(PatchCollection $collection, PackageEvent $event)
                 $extra = $package->getExtra();
                 if (isset($extra['patches'])) {
                     $patches = $this->findPatchesInJson($extra['patches']);
-                    foreach ($patches as $package => $patch_list) {
+                    foreach ($patches as $patch_package => $patch_list) {
```
Comment (EmiiKhaos) [inline src/Resolvers/DependencyPatches.php]: Code style: `$patchPackage`, `$patchList`

**Comment by drupol** [comment]
@rlmumford any update on this ? It would be nice to finalize this nice feature

**Comment by drupol** [comment]
See #247 for an updated version of this PR.

It fixes the coding style and add a new test.

**Comment by cweagans** [comment]
Resolving patches from dependencies isn't a thing in 2.x anymore.

**Pull request closed without merging**